File: src/main/java/org/lwjgl/opengl/Display.java
package org.lwjgl.opengl;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;

import net.minecraft.client.render.window.GameWindowGLFW;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.lwjgl.BufferUtils;
import org.lwjgl.LWJGLException;
import org.lwjgl.glfw.GLFW;
import org.lwjgl.glfw.GLFWErrorCallback;
import org.lwjgl.glfw.GLFWImage;
import org.lwjgl.glfw.GLFWVidMode;
import org.lwjgl.glfw.GLFWWindowSizeCallback;
import org.lwjgl.input.Keyboard;
import org.lwjgl.input.Mouse;
import org.lwjgl.system.MemoryUtil;

@SuppressWarnings("unused")
public final class Display {
    @NotNull
    public static final Display INSTANCE = new Display();
    @NotNull
    private static String title = "";
    @NotNull
    private static DisplayMode displayMode = new DisplayMode(640, 480, 24, 60);
    private static long handleKt = -1L;
    private static boolean resizable;
    private static int widthKt;
    private static int heightKt;
    private static int xPos;
    private static int yPos;
    @Nullable
    private static GameWindowGLFW window;
    private static boolean window_resized;
    @Nullable
    private static GLFWWindowSizeCallback sizeCallback;
    private static ByteBuffer @Nullable [] cached_icons = null;

    private Display() {
    }

    @NotNull
    @SuppressWarnings("unused")
    public static String getTitle() {
        return title;
    }

    @SuppressWarnings("unused")
    public static void setTitle(@NotNull String value) {
        Objects.requireNonNull(value);
        title = value;
        if (isCreatedKt()) {
            GLFW.glfwSetWindowTitle(handleKt, value);
        }
    }

    @NotNull
    @SuppressWarnings("unused")
    public static DisplayMode getDisplayMode() {
        return displayMode;
    }

    @SuppressWarnings("unused")
    public static void setDisplayMode(@NotNull DisplayMode displayMode) {
        Objects.requireNonNull(displayMode);
        Display.displayMode = displayMode;
    }

    @SuppressWarnings("unused")
    public static long getHandleKt() {
        return handleKt;
    }

    @SuppressWarnings("unused")
    public static void setHandleKt(long handle) {
        handleKt = handle;
    }

    @SuppressWarnings("unused")
    public static int getWidthKt() {
        return widthKt;
    }

    @SuppressWarnings("unused")
    public static void setWidthKt(int width) {
        widthKt = width;
    }

    @SuppressWarnings("unused")
    public static int getHeightKt() {
        return heightKt;
    }

    @SuppressWarnings("unused")
    public static void setHeightKt(int height) {
        heightKt = height;
    }

    @SuppressWarnings("unused")
    public static int getXPos() {
        return xPos;
    }

    @SuppressWarnings("unused")
    public static void setXPos(int xPos) {
        Display.xPos = xPos;
    }

    @SuppressWarnings("unused")
    public static int getYPos() {
        return yPos;
    }

    @SuppressWarnings("unused")
    public static void setYPos(int yPos) {
        Display.yPos = yPos;
    }

    @Nullable
    @SuppressWarnings("unused")
    public static GameWindowGLFW getWindow() {
        return window;
    }

    @SuppressWarnings("unused")
    public static void setWindow(@Nullable GameWindowGLFW value) {
        window = value;
    }

    @Nullable
    @SuppressWarnings("unused")
    public static DisplayMode getDesktopDisplayMode() {
        DisplayMode[] $this$maxByOrNull$iv = getAvailableDisplayModes();
        DisplayMode var12;
        if ($this$maxByOrNull$iv.length == 0) {
            var12 = null;
        } else {
            DisplayMode maxElem$iv = $this$maxByOrNull$iv[0];
            int lastIndex$iv = ($this$maxByOrNull$iv.length - 1);
            if (lastIndex$iv != 0) {
                DisplayMode it = maxElem$iv;
                int maxValue$iv = it.getWidth() * it.getHeight();
                int i$iv = 1;
                if (i$iv <= lastIndex$iv) {
                    while (true) {
                        DisplayMode e$iv = $this$maxByOrNull$iv[i$iv];
                        int v$iv = it.getWidth() * it.getHeight();
                        if (maxValue$iv < v$iv) {
                            maxElem$iv = e$iv;
                            maxValue$iv = v$iv;
                        }

                        if (i$iv == lastIndex$iv) {
                            break;
                        }

                        ++i$iv;
                    }
                }

            }
            var12 = maxElem$iv;
        }

        return var12;
    }

    @SuppressWarnings({"unused", "UnusedReturnValue"})
    public static int setIcon(@NotNull ByteBuffer[] icons) {
        Objects.requireNonNull(icons);
        if (!Arrays.equals(cached_icons, icons)) {
            Collection<ByteBuffer> destination$iv$iv = new ArrayList<>(icons.length);
            int var6 = 0;

            for(int var7 = icons.length; var6 < var7; ++var6) {
                ByteBuffer it = icons[var6];
                ByteBuffer var12 = INSTANCE.cloneByteBuffer(it);
                destination$iv$iv.add(var12);
            }

            cached_icons = destination$iv$iv.toArray(new ByteBuffer[0]);
        }

        if (isCreatedKt()) {
            GLFW.glfwSetWindowIcon(handleKt, INSTANCE.iconsToGLFWBuffer(Objects.requireNonNull(cached_icons)));
            return 1;
        } else {
            return 0;
        }
    }

    private ByteBuffer cloneByteBuffer(ByteBuffer original) {
        ByteBuffer clone = BufferUtils.createByteBuffer(original.capacity());
        int old_position = original.position();
        clone.put(original);
        original.position(old_position);
        clone.flip();
        return clone;
    }

    private GLFWImage.Buffer iconsToGLFWBuffer(ByteBuffer[] icons) {
        GLFWImage.Buffer buffer = GLFWImage.create(icons.length);

        for (ByteBuffer icon : icons) {
            int size = icon.limit() / 4;
            int dimension = (int)Math.sqrt(size);
            try (GLFWImage image = GLFWImage.malloc()) {
                buffer.put(image.set(dimension, dimension, icon));
            }
        }

        buffer.flip();
        return buffer;
    }

    @SuppressWarnings("unused")
    public static void update() {
        window_resized = false;
        GLFW.glfwPollEvents();
        if (Mouse.isCreated()) {
            Mouse.poll();
//            Mouse.updateCursor();
        }

        if (Keyboard.isCreated()) {
            Keyboard.poll();
        }

        GLFW.glfwSwapBuffers(handleKt);
    }

    @SuppressWarnings("unused")
    public static void create(@Nullable PixelFormat pixelFormat) throws LWJGLException {
        GLFWErrorCallback.createPrint(System.err).set();
        boolean var1 = GLFW.glfwInit();
        if (!var1) {
            String var3 = "Unable to initialize GLFW";
            throw new IllegalStateException(var3);
        }
        GLFW.glfwDefaultWindowHints();
        GLFW.glfwWindowHint(GLFW.GLFW_VISIBLE, GLFW.GLFW_FALSE);
        GLFW.glfwWindowHint(GLFW.GLFW_RESIZABLE, resizable ? GLFW.GLFW_TRUE : GLFW.GLFW_FALSE);
        handleKt = GLFW.glfwCreateWindow(displayMode.getWidth(), displayMode.getHeight(), title,MemoryUtil.NULL, MemoryUtil.NULL);
        widthKt = displayMode.getWidth();
        heightKt = displayMode.getHeight();
        GLFW.glfwMakeContextCurrent(handleKt);
        GL.createCapabilities();
        sizeCallback = GLFWWindowSizeCallback.create(INSTANCE::resizeCallback);
        GLFW.glfwSetWindowSizeCallback(handleKt, sizeCallback);
        Mouse.create();
        Keyboard.create();
        GLFW.glfwShowWindow(handleKt);
        if (cached_icons != null) {
            setIcon(cached_icons);
        }
    }

    @SuppressWarnings("unused")
    public static void setFullscreen(boolean fullscreen) {
        System.out.println("setFullscreen: " + fullscreen);

        try {
            INSTANCE.resizeCallback(handleKt, displayMode.getWidth(), displayMode.getHeight());
            if (fullscreen) {
                long monitor = GLFW.glfwGetPrimaryMonitor();
                GLFW.glfwSetWindowMonitor(handleKt, monitor, 0, 0, widthKt, heightKt, displayMode.getFrequency());
                xPos = displayMode.getWidth() / 2;
                yPos = displayMode.getHeight() / 2;
            } else {
                xPos -= widthKt / 2;
                yPos -= heightKt / 2;
                GLFW.glfwSetWindowMonitor(handleKt, 0L, xPos, yPos, widthKt, heightKt, -1);
            }

            GLFW.glfwSetWindowSize(handleKt, widthKt, heightKt);
        } catch (Throwable t) {
            t.printStackTrace();
        }
    }

    @NotNull
    public static DisplayMode[] getAvailableDisplayModes() {
        long primaryMonitor = GLFW.glfwGetPrimaryMonitor();
        if (primaryMonitor == MemoryUtil.NULL) {
            return new DisplayMode[0];
        } else {
            GLFWVidMode.Buffer modes = GLFW.glfwGetVideoModes(primaryMonitor);
            if (modes == null) {
                throw new IllegalStateException("No video modes found");
            } else {
                List<DisplayMode> displayModes = new ArrayList<>();

                for (GLFWVidMode mode : modes) {
                    displayModes.add(new DisplayMode(mode.width(), mode.height(), mode.redBits() + mode.blueBits() + mode.greenBits(), mode.refreshRate()));
                }

                return new HashSet<>(displayModes).toArray(new DisplayMode[0]);
            }
        }
    }

    private void resizeCallback(long window, int width, int height) {
        if (window == handleKt) {
            window_resized = true;
            widthKt = width;
            heightKt = height;
        }

    }

    @SuppressWarnings("unused")
    public void destroyWindow() {
        // free callbacks
        Objects.requireNonNull(sizeCallback).free();
        Mouse.destroy();
        Keyboard.destroy();
        // Destroy the window
        GLFW.glfwDestroyWindow(handleKt);
    }

    @SuppressWarnings("unused")
    public static long getHandle() {
        return Objects.requireNonNull(getWindow()).window;
    }

    @SuppressWarnings("unused")
    public static int getWidth() {
        return Objects.requireNonNull(getWindow()).getWidthScreenCoords();
    }

    @SuppressWarnings("unused")
    public static int getHeight() {
        return Objects.requireNonNull(getWindow()).getHeightScreenCoords();
    }

    @SuppressWarnings("unused")
    public static boolean isCreated() {
        return Objects.requireNonNull(getWindow()).window != -1L;
    }

    @SuppressWarnings("unused")
    public static void destroy() {
        INSTANCE.destroyWindow();
        GLFW.glfwTerminate();
        GLFWErrorCallback callback = GLFW.glfwSetErrorCallback(null);
        if (callback != null) {
            callback.close();
        }
    }

    @SuppressWarnings("unused")
    public static boolean isCreatedKt() {
        return handleKt != -1L;
    }

    @SuppressWarnings("unused")
    public static boolean isCloseRequested() {
        return GLFW.glfwWindowShouldClose(handleKt);
    }

    @SuppressWarnings("unused")
    public static boolean isActive() {
        return true;
    }

    @SuppressWarnings("unused")
    public static void setResizable(boolean isResizable) {
        resizable = isResizable;
        if (isCreatedKt()) {
            GLFW.glfwWindowHint(GLFW.GLFW_RESIZABLE, resizable ? GLFW.GLFW_TRUE : GLFW.GLFW_FALSE);
        }
    }

    @SuppressWarnings("unused")
    public static void sync(int fps) {
        Sync.sync(fps);
    }

    @SuppressWarnings("unused")
    public static void setVSyncEnabled(boolean enabled) {
        GLFW.glfwSwapInterval(enabled ? GLFW.GLFW_TRUE : GLFW.GLFW_FALSE);
    }

    @SuppressWarnings("unused")
    public static boolean wasResized() {
        return window_resized;
    }
}

File: src/main/java/org/lwjgl/opengl/Display.kt
package org.lwjgl.opengl

import net.minecraft.client.render.window.GameWindowGLFW
import org.lwjgl.BufferUtils
import org.lwjgl.glfw.GLFW
import org.lwjgl.glfw.GLFWErrorCallback
import org.lwjgl.glfw.GLFWImage
import org.lwjgl.glfw.GLFWWindowSizeCallback
import org.lwjgl.input.Keyboard
import org.lwjgl.input.Mouse
import org.lwjgl.system.MemoryUtil
import java.nio.Buffer
import java.nio.ByteBuffer
import kotlin.math.sqrt


object Display {
    @JvmStatic
    var title: String = ""
        set(value) {
            field = value
            if (isCreatedKt) {
                GLFW.glfwSetWindowTitle(handleKt, value)
            }
        }

    @JvmStatic
    var displayMode: DisplayMode = DisplayMode(640, 480, 24, 60)

    @JvmStatic
    var handleKt: Long = -1
    private var resizable = false

    @JvmStatic
    var widthKt = 0

    @JvmStatic
    var heightKt = 0

    @JvmStatic
    var xPos = 0

    @JvmStatic
    var yPos = 0

    @JvmStatic
    var window: GameWindowGLFW? = null
	    get() = field
	    set(value) {
		    field = value
	    }

	@JvmStatic
    val desktopDisplayMode: DisplayMode?
     get() { return availableDisplayModes.maxByOrNull { it.width * it.height } }
    private var window_resized = false
    private var sizeCallback: GLFWWindowSizeCallback? = null

    private var cached_icons: Array<ByteBuffer>? = null
    @JvmStatic
    fun setIcon(icons: Array<ByteBuffer>): Int {
        // LWJGL2 doesn't enforce this to be called after window creation,
        // meaning you have to keep hold the icons to use them when the window is created
        if (!this.cached_icons.contentEquals(icons)) {
            // you have to also clone the byte buffers to avoid seg faults from them being freed
            cached_icons = icons.map { cloneByteBuffer(it) }.toTypedArray()
        }

        if (this.isCreatedKt) {
            GLFW.glfwSetWindowIcon(this.handleKt, iconsToGLFWBuffer(this.cached_icons!!))
            return 1
        } else {
            return 0
        }
    }


    private fun cloneByteBuffer(original: ByteBuffer): ByteBuffer {
        // code taken from LWJGL2' Display.java
        val clone = BufferUtils.createByteBuffer(original.capacity())
        val old_position = original.position()
        clone.put(original)
        // if compiled with java 11+ calls ByteBuffer.position/ByteBuffer.flip however java <= 8 doesnt have these
        // methods and instead compiles as calls to Buffer.position/Buffer.flip causing crashes on those java versions
        (original as Buffer).position(old_position)
        (clone as Buffer).flip()

        return clone
    }
    private fun iconsToGLFWBuffer(icons: Array<ByteBuffer>): GLFWImage.Buffer {
        val buffer = GLFWImage.create(icons.size)
        icons.forEach { icon ->
            val size: Int = icon.limit() / 4
            val dimension = sqrt(size.toDouble()).toInt()
            GLFWImage.malloc().use { image ->
                buffer.put(image.set(dimension, dimension, icon))
            }
        }
        buffer.flip()
        return buffer
    }

    @JvmStatic
    fun update() {
        window_resized = false
        GLFW.glfwPollEvents()
        if ( Mouse.isCreated() ) {
            Mouse.poll()
//            Mouse.updateCursor();
        }

        if ( Keyboard.isCreated() ) {
            Keyboard.poll()
        }
        GLFW.glfwSwapBuffers(handleKt)
    }

    @JvmStatic
    fun create(pixelFormat: PixelFormat?) {
        // Setup an error callback. The default implementation
        GLFWErrorCallback.createPrint(System.err).set()
        check(GLFW.glfwInit()) { "Unable to initialize GLFW" }
        // Configure GLFW
        GLFW.glfwDefaultWindowHints() // optional, the current window hints are already the default
        GLFW.glfwWindowHint(GLFW.GLFW_VISIBLE, GLFW.GLFW_FALSE) // the window will stay hidden after creation
        GLFW.glfwWindowHint(GLFW.GLFW_RESIZABLE, if (resizable) GLFW.GLFW_TRUE else GLFW.GLFW_FALSE)
        handleKt =
            GLFW.glfwCreateWindow(displayMode.width, displayMode.height, title, MemoryUtil.NULL, MemoryUtil.NULL)
        widthKt = displayMode.width
        heightKt = displayMode.height
        GLFW.glfwMakeContextCurrent(handleKt)
        GL.createCapabilities()
        // create general callbacks
        sizeCallback = GLFWWindowSizeCallback.create(Display::resizeCallback)
        GLFW.glfwSetWindowSizeCallback(handleKt, sizeCallback)
        Mouse.create()
        Keyboard.create()
        GLFW.glfwShowWindow(handleKt)
        if (this.cached_icons != null) {
            this.setIcon(this.cached_icons!!)
        }
    }

    /**
     * This is a shit code snippet, but at least it works.
     */
    @JvmStatic
    fun setFullscreen(fullscreen: Boolean) {
        println("setFullscreen: $fullscreen")
        runCatching {
            this.resizeCallback(handleKt, displayMode.width, displayMode.height)
            if (fullscreen) {
                var monitor = GLFW.glfwGetPrimaryMonitor()

                GLFW.glfwSetWindowMonitor(
                    handleKt,
                    monitor,
                    0,
                    0,
                    widthKt,
                    heightKt,
                    displayMode.frequency
                )
                xPos = displayMode.width / 2
                yPos = displayMode.height / 2
            } else {
                xPos -= widthKt / 2
                yPos -= heightKt / 2
                GLFW.glfwSetWindowMonitor(
                    handleKt,
                    0L,
                    xPos,// need a xPos
                    yPos,// need a yPos
                    widthKt,
                    heightKt,
                    -1
                )
            }
            GLFW.glfwSetWindowSize(handleKt, widthKt, heightKt)
        }.onFailure {
            it.printStackTrace()
        }
    }
    @JvmStatic
    val availableDisplayModes: Array<DisplayMode>
        get() {
            val primaryMonitor = GLFW.glfwGetPrimaryMonitor()
            if (primaryMonitor == MemoryUtil.NULL) {
                return arrayOf()
            }
            val videoModes = GLFW.glfwGetVideoModes(primaryMonitor) ?: error("No video modes found")
            return videoModes.map { mode ->
                DisplayMode(
                    mode.width(),
                    mode.height(),
                    mode.redBits() + mode.blueBits() + mode.greenBits(),
                    mode.refreshRate()
                )
            }.toHashSet().toTypedArray()
        }

    private fun resizeCallback(window: Long, width: Int, height: Int) {
        if (window == handleKt) {
            window_resized = true
            Display.widthKt = width
            Display.heightKt = height
        }
    }

    fun destroyWindow() {
        // free callbacks
        sizeCallback!!.free()
        Mouse.destroy()
        Keyboard.destroy()
        // Destroy the window
        GLFW.glfwDestroyWindow(handleKt)
    }

    @JvmStatic
    fun getHandle(): Long {
        return window!!.window
    }

    @JvmStatic
    fun getWidth(): Int {
        return window!!.getWidthScreenCoords()
    }

    @JvmStatic
    fun getHeight(): Int {
        return window!!.getHeightScreenCoords()
    }

    @JvmStatic
    fun isCreated(): Boolean {
        return window!!.window != -1L
    }

    @JvmStatic
    fun destroy() {
        destroyWindow()
        // Terminate GLFW and free the error callback
        GLFW.glfwTerminate()
        GLFW.glfwSetErrorCallback(null)?.free()
    }

    @JvmStatic
    val isCreatedKt: Boolean
        get() = handleKt != -1L

    @JvmStatic
    val isCloseRequested: Boolean
        get() = GLFW.glfwWindowShouldClose(handleKt)

    @JvmStatic
    val isActive: Boolean
        get() = true

    @JvmStatic
    fun setResizable(isResizable: Boolean) {
        resizable = isResizable
        if (isCreatedKt) {
            GLFW.glfwWindowHint(GLFW.GLFW_RESIZABLE, if (resizable) GLFW.GLFW_TRUE else GLFW.GLFW_FALSE)
        }
    }

    @JvmStatic
    fun sync(fps: Int) {
        Sync.sync(fps)
    }

    @JvmStatic
    fun setVSyncEnabled(enabled: Boolean) {
        GLFW.glfwSwapInterval(if (enabled) 1 else 0)
    }

    @JvmStatic
    fun wasResized(): Boolean = window_resized
}